core/modules/layout/layout.api.php

```diff
@@ -196,6 +196,191 @@ function hook_layout_renderer_info() {
 }
 
 /**
+ * Respond to a block being removed from a layout.
+ *
+ * This hook is invoked from Layout::removeBlock() after a block has been 
+ * removed from a layout.
+ * In the context of the Layout Builder User Interface, this hook is 
+ * executed when a user clicks the "Remove block" link in the dropbutton
+ * which displays the block edit options.
+ * In a custom module, if custom settings have been saved as a result of
+ * a hook_block_save() operation, hook_layout_block_remove() allows
+ * these settings to be deleted if the block is removed before the layout
+ * is saved. 
+ *
+ * @param Layout $layout
+ *   The layout object from which the block is being removed.
+ * @param Block $block
+ *   The block object which is being removed.
+ *
+ * @see hook_block_save()
+ * @see hook_block_add()
+ *
+ * @ingroup layout_api_hooks
+ */
+function hook_layout_block_remove(Block $block, Layout $layout) {
+  if ($block->delta == 'my_block') {
+    config_set('my_module', 'custom_settings', array());
+  }
+}
+
+/**
+ * Respond to a block being programmatically added to a layout.
+ *
+ * This hook is invoked from Layout::addBlock() after a block has been 
+ * programmatically added to a region in a layout.
+ * In the context of the Layout Builder User Interface, this hook does NOT run
+ * after a user has clicked the "Add Block" button to insert the block into a 
+ * region.
```

Comment:
Could you explain when this *does* fire in the context of the UI? When the layout is saved? And if this is the case, wouldn't it make more sense for `hook_layout_block_remove()` fire on save as well? These docs are great, but they are pointing out some discrepancies in the calling of hooks which seem like they should be made consistent rather than just explained.